File: packages/ts/src/config/validate.ts
import {
  DA_BACKENDS,
  GAS_MODELS,
  PROOF_SYSTEMS,
  SEQUENCER_MODES,
  SETTLEMENT_PARADIGMS,
  VM_TYPES,
} from "./enums";
import { isProofCompatible, requiresBasedSequencer, validProofSystems } from "./matrix";
import type { RollupConfig } from "./types";
import { RollupConfigError } from "./errors";

/** The outcome of validating a {@link RollupConfig}. */
export interface ValidationResult {
  /** True when there are no errors (warnings do not affect validity). */
  valid: boolean;
  /** Hard failures that block submission. */
  errors: string[];
  /** Non-fatal notices (e.g. selecting a not-yet-active DA backend). */
  warnings: string[];
}

function isPositiveIntegerString(value: string): boolean {
  return /^[1-9][0-9]*$/.test(value);
}

/**
 * Validate a rollup configuration against the on-chain rules: the
 * settlement → proof compatibility matrix, the based-settlement ⇒ based-sequencer
 * constraint, the closed value sets, and basic field sanity.
 *
 * Returns a structured result; callers that prefer to fail fast can use
 * {@link assertValidRollupConfig}.
 */
export function validateRollupConfig(config: RollupConfig): ValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  if (!config.rollupId || config.rollupId.trim() === "") {
    errors.push("rollupId must be a non-empty string");
  }

  if (!SETTLEMENT_PARADIGMS.includes(config.settlement)) {
    errors.push(`settlement "${config.settlement}" is not a valid settlement paradigm`);
  }
  if (!SEQUENCER_MODES.includes(config.sequencer)) {
    errors.push(`sequencer "${config.sequencer}" is not a valid sequencer mode`);
  }
  if (!PROOF_SYSTEMS.includes(config.proofSystem)) {
    errors.push(`proofSystem "${config.proofSystem}" is not a valid proof system`);
  }
  if (!DA_BACKENDS.includes(config.da)) {
    errors.push(`da "${config.da}" is not a valid data-availability backend`);
  }
  if (!GAS_MODELS.includes(config.gasModel)) {
    errors.push(`gasModel "${config.gasModel}" is not a valid gas model`);
  }
  if (!VM_TYPES.includes(config.vmType)) {
    errors.push(`vmType "${config.vmType}" is not a valid VM type`);
  }

  // Compatibility matrix (only meaningful once both values are valid).
  if (
    SETTLEMENT_PARADIGMS.includes(config.settlement) &&
    PROOF_SYSTEMS.includes(config.proofSystem) &&
    !isProofCompatible(config.settlement, config.proofSystem)
  ) {
    errors.push(
      `proof system "${config.proofSystem}" is not compatible with "${config.settlement}" ` +
        `settlement (expected one of: ${validProofSystems(config.settlement).join(", ")})`,
    );
  }

  // Based settlement requires the based sequencer mode.
  if (requiresBasedSequencer(config.settlement) && config.sequencer !== "based") {
    errors.push('based settlement requires the "based" sequencer mode');
  }

  if (!Number.isInteger(config.blockTimeMs) || config.blockTimeMs <= 0) {
    errors.push("blockTimeMs must be a positive integer");
  }
  if (!Number.isInteger(config.maxTxPerBlock) || config.maxTxPerBlock <= 0) {
    errors.push("maxTxPerBlock must be a positive integer");
  }

  if (config.stakeAmountUqor !== undefined && !isPositiveIntegerString(config.stakeAmountUqor)) {
    errors.push("stakeAmountUqor must be a positive integer string (base uqor)");
  }
  if (
    config.challengeWindowSecs !== undefined &&
    (!Number.isInteger(config.challengeWindowSecs) || config.challengeWindowSecs <= 0)
  ) {
    errors.push("challengeWindowSecs must be a positive integer");
  }
  if (
    config.maxDaBlobSize !== undefined &&
    (!Number.isInteger(config.maxDaBlobSize) || config.maxDaBlobSize <= 0)
  ) {
    errors.push("maxDaBlobSize must be a positive integer (bytes)");
  }

  // Celestia is a selectable but not-yet-active backend on the network.
  if (config.da === "celestia" || config.da === "both") {
    warnings.push(
      "Celestia data availability is selectable but not yet active on the network; " +
        "batches targeting it will not be served until it is enabled.",
    );
  }

  return { valid: errors.length === 0, errors, warnings };
}

/** Validate a configuration and throw {@link RollupConfigError} on any error. */
export function assertValidRollupConfig(config: RollupConfig): void {
  const result = validateRollupConfig(config);
  if (!result.valid) {
    throw new RollupConfigError(result.errors);
  }
}

File: packages/rust/src/config/validate.rs
//! Client-side validation of a rollup configuration against the on-chain rules.

use super::enums::{DaBackend, Sequencer};
use super::errors::RollupConfigError;
use super::matrix::{is_proof_compatible, requires_based_sequencer, valid_proof_systems};
use super::types::RollupConfig;

/// The outcome of validating a [`RollupConfig`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ValidationResult {
    /// True when there are no errors (warnings do not affect validity).
    pub valid: bool,
    /// Hard failures that block submission.
    pub errors: Vec<String>,
    /// Non-fatal notices (e.g. selecting a not-yet-active DA backend).
    pub warnings: Vec<String>,
}

fn is_positive_integer_string(value: &str) -> bool {
    !value.is_empty()
        && value.bytes().next() != Some(b'0')
        && value.bytes().all(|b| b.is_ascii_digit())
}

/// Validate a rollup configuration against the on-chain rules: the
/// settlement -> proof compatibility matrix, the based-settlement =>
/// based-sequencer constraint, and basic field sanity.
///
/// Returns a structured result; callers that prefer to fail fast can use
/// [`assert_valid_rollup_config`].
pub fn validate_rollup_config(config: &RollupConfig) -> ValidationResult {
    let mut errors: Vec<String> = Vec::new();
    let mut warnings: Vec<String> = Vec::new();

    if config.rollup_id.trim().is_empty() {
        errors.push("rollupId must be a non-empty string".to_string());
    }

    // The strongly typed enums guarantee valid value sets; the matrix and the
    // based-sequencer constraint are the remaining cross-field rules.
    if !is_proof_compatible(config.settlement, config.proof_system) {
        let allowed: Vec<&str> = valid_proof_systems(config.settlement)
            .iter()
            .map(|p| p.as_str())
            .collect();
        errors.push(format!(
            "proof system \"{}\" is not compatible with \"{}\" settlement (expected one of: {})",
            config.proof_system.as_str(),
            config.settlement.as_str(),
            allowed.join(", "),
        ));
    }

    if requires_based_sequencer(config.settlement) && config.sequencer != Sequencer::Based {
        errors.push("based settlement requires the \"based\" sequencer mode".to_string());
    }

    if config.block_time_ms <= 0 {
        errors.push("blockTimeMs must be a positive integer".to_string());
    }
    if config.max_tx_per_block <= 0 {
        errors.push("maxTxPerBlock must be a positive integer".to_string());
    }

    if let Some(stake) = &config.stake_amount_uqor {
        if !is_positive_integer_string(stake) {
            errors
                .push("stakeAmountUqor must be a positive integer string (base uqor)".to_string());
        }
    }
    if let Some(window) = config.challenge_window_secs {
        if window <= 0 {
            errors.push("challengeWindowSecs must be a positive integer".to_string());
        }
    }
    if let Some(blob) = config.max_da_blob_size {
        if blob <= 0 {
            errors.push("maxDaBlobSize must be a positive integer (bytes)".to_string());
        }
    }

    if config.da == DaBackend::Celestia || config.da == DaBackend::Both {
        warnings.push(
            "Celestia data availability is selectable but not yet active on the network; \
             batches targeting it will not be served until it is enabled."
                .to_string(),
        );
    }

    ValidationResult {
        valid: errors.is_empty(),
        errors,
        warnings,
    }
}

/// Validate a configuration and return [`RollupConfigError`] on any error.
pub fn assert_valid_rollup_config(config: &RollupConfig) -> Result<(), RollupConfigError> {
    let result = validate_rollup_config(config);
    if result.valid {
        Ok(())
    } else {
        Err(RollupConfigError::new(result.errors))
    }
}
